src/pymor/reductors/era.py

```diff
@@ -0,0 +1,238 @@
+# This file is part of the pyMOR project (https://www.pymor.org).
+# Copyright pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
+
+import numpy as np
+import scipy.linalg as spla
+
+from pymor.algorithms.projection import project
+from pymor.algorithms.to_matrix import to_matrix
+from pymor.core.cache import cached, CacheableObject
+from pymor.models.iosys import LTIModel
+from pymor.operators.interface import Operator
+from pymor.operators.numpy import NumpyHankelOperator, NumpyMatrixOperator
+
+
+class ERAReductor(CacheableObject):
+    r"""Eigensystem Realization Algorithm reductor.
+
+    Constructs a (reduced) realization from a sequence of Markov parameters :math:`h_i`,
+    for :math:`i\in\{1,\,\dots,\,2s-1\}`, :math:`s\in\mathbb{N}`, by a (reduced) orthogonal
+    factorization of the Hankel matrix of Markov parameters
+
+    .. math::
+        H =
+        \begin{bmatrix}
+            h_1 & h_2 & \dots & h_s \\
+            h_2 & h_3 & \dots & h_{s+1}\\
+            \vdots & \vdots && \vdots\\
+            h_s & h_{s+1} & \dots & h_{2s-1}
+        \end{bmatrix}=U\Sigma V^T\in\mathbb{R}^{ms\times ps},
+
+    where :math:`r\leq\min\{ms,ps\}` is the reduced order. See :cite:`K78`.
+
+    In order for the identified model to be stable, the Markov parameters must satisfy
+
+    .. math::
+        h_i\rightarrow0~\text{for}~i>s.
```

Comment:
This doesn't look very precise. Is it supposed to represent the limit as $i \to \infty$? For a finite amount of data, that doesn't seem to be useful.

Reply:
I also think that this criterion is rather fuzzy.

I have encountered this equation a couple of times in the literature, like [Boris Kramer's TERA paper](https://www.tandfonline.com/doi/full/10.1080/13873954.2016.1198389), where it is  cited from Kungs original work.

Despite the paper having [over 800 citations](https://scholar.google.com/scholar?cluster=13487788052018306962&hl=en&as_sdt=0,5), I have been unsuccessfully looking for a copy for more than three years now...

Reply:
> Is it supposed to represent the limit as i→∞?

It is supposed to mean that the sequence of $2s-1$ Markov parameters should have decayed 'significantly' after $s$ steps.

Reply:
> Despite the paper having [over 800 citations](https://scholar.google.com/scholar?cluster=13487788052018306962&hl=en&as_sdt=0,5), I have been unsuccessfully looking for a copy for more than three years now...

Fun...
I had no more luck than you.

> It is supposed to mean that the sequence of 2s−1 Markov parameters should have decayed 'significantly' after s steps.

I would then be for removing the "mathematical" condition and only saying it in words.